File: extensions/Presentation/Presentation/Presentation.js
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

var CHANGE_EVENT_NAME = 'change';
var DEFAULT_SESSION_START_EVENT_NAME = 'defaultsessionstart';

var _presentationSessions = {};
var _availabilityList = [];

function Availability() {
    this.value = false;
    this.__defineSetter__("on" + CHANGE_EVENT_NAME, function(callback) {
        if (callback) {
            addEventListener(CHANGE_EVENT_NAME, callback);
        } else {
            removeEventListener(CHANGE_EVENT_NAME, this.onchange);
        }
    });
}

function randomPresentationId() {
    return Math.random().toString(18).slice(2);
}

exports.__defineSetter__("on" + DEFAULT_SESSION_START_EVENT_NAME, function(callback) {
     if (callback) {
         addEventListener(DEFAULT_SESSION_START_EVENT_NAME, callback);
     } else {
         removeEventListener(DEFAULT_SESSION_START_EVENT_NAME, this.ondefaultsessionstart);
     }
});

exports.startSession = function(url, presentationId) {
    var _this = this;

    return new Promise(function(resolve, reject) {
        if (presentationId == undefined) {
            presentationId = randomPresentationId();
            while (_presentationSessions[presentationId] != undefined) {
                presentationId = randomPresentationId();
            }
        } else if (_presentationSessions[presentationId] != undefined) {
            throw new Error("Duplicate Presentation ID");
            reject();
        }

        var presentation = new navigator.presentation.PresentationSession(url, presentationId);
        _presentationSessions[presentationId] = presentation;

        _this.invokeNative('startSession', [url, presentationId, {'resolve': resolve, 'reject': reject}]);

        resolve(presentation);
    });
}

exports.joinSession = function(url, presentationId) {
    return new Promise(function(resolve, reject) {
    });
}

exports.getAvailability = function() {
    var _this = this;
    return new Promise(function(resolve, reject) {
        var resolveWrapper = function(available) {
            var availability = new Availability();
            availability.value = available;
            resolve(availability);
        }
        _this.invokeNative('getAvailability', [{'resolve': resolveWrapper, 'reject': reject}]);
    });
}

File: extensions/Presentation/Presentation/Presentation.swift
// Copyright (c) 2015 Intel Corporation. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import Foundation
import XWalkView

public class Presentation: XWalkExtension {

    var remoteWindow: UIWindow? = nil
    var remoteViewController: RemoteViewController? = nil
    var sessions: Dictionary<String, PresentationSessionHost> = [:]

    override init() {
        super.init()
        if UIScreen.screens().count == 2 {
            print("external display connected. count:\(UIScreen.screens().count)")
            let remoteScreen: UIScreen = UIScreen.screens()[1];
            createWindowForScreen(remoteScreen)
            sendAvailableChangeEvent(true)
        }

        NSNotificationCenter.defaultCenter().addObserver(self,
            selector: "screenDidConnect:", name: UIScreenDidConnectNotification, object: nil)
        NSNotificationCenter.defaultCenter().addObserver(self,
            selector: "screenDidDisconnect:", name: UIScreenDidDisconnectNotification, object: nil)
    }

    public override func didBindExtension(channel: XWalkChannel!, instance: Int) {
        super.didBindExtension(channel, instance: instance)

        let extensionName = "navigator.presentation.PresentationSession"
        PresentationSessionHost.presentation = self
        if let ext: PresentationSessionHost = XWalkExtensionFactory.createExtension(extensionName) as? PresentationSessionHost {
            channel.webView?.loadExtension(ext, namespace: extensionName)
        }
    }

    func createWindowForScreen(screen: UIScreen) {
        assert(remoteWindow == nil, "remoteWindow should be nil before it is created.")
        remoteWindow = UIWindow(frame: screen.bounds)
        remoteWindow?.screen = screen

        remoteViewController = RemoteViewController()
        remoteWindow?.rootViewController = remoteViewController
        remoteWindow?.hidden = false
    }

    func sendAvailableChangeEvent(isAvailable: Bool) {
        var js = "var ev = new Event(\"change\"); ev.value = \(isAvailable);";
        js += " dispatchEvent(ev);";
        self.evaluateJavaScript(js)
    }

    func screenDidConnect(notification: NSNotification) {
        if let screen = notification.object as? UIScreen {
            createWindowForScreen(screen)
            sendAvailableChangeEvent(true)
        }
    }

    func screenDidDisconnect(notification: NSNotification) {
        if let _ = notification.object as? UIScreen {
            sendAvailableChangeEvent(false)
            remoteWindow?.hidden = true
            remoteViewController?.willDisconnect()
            remoteViewController = nil
            remoteWindow = nil
        }
    }

    func registerSession(sessionHost: PresentationSessionHost, id: String) {
        sessions[id] = sessionHost
    }

    func unregisterSession(id: String) {
        sessions[id] = nil
    }

    func jsfunc_startSession(cid: UInt32, url: String, presentationId: String, _Promise: UInt32) {
        if let sessionHost = sessions[presentationId] {
            remoteViewController?.attachSession(sessionHost)
        }
    }

    func jsfunc_joinSession(cid: UInt32, url: String, presentationId: String, _Promise: UInt32) {
    }

    func jsfunc_getAvailability(cid: UInt32, _Promise: UInt32) {
        invokeCallback(_Promise, key: "resolve", arguments: [UIScreen.screens().count == 2]);
    }
}
